c/examples/cpp-experiment.cpp

```diff
@@ -0,0 +1,215 @@
+/* C++ wrapper experiement. */
+
+#include <iostream>
+#include <stdexcept>
+#include <sstream>
+#include <memory>
+#include <cassert>
+
+#include <tskit/tables.h>
+
+using namespace std;
+
+void
+check_error(int val)
+{
+    if (val < 0) {
+        std::ostringstream o;
+        o << tsk_strerror(val);
+        throw std::runtime_error(o.str());
+    }
+}
+
+namespace tskit {
+
+    class NodeTable
+    {
+
+        private:
+            tsk_node_table_t *table;
+            const bool allocated_locally;
+            using node_table_ptr = std::unique_ptr<tsk_node_table_t,void(*)(tsk_node_table_t*)>;
+
+            node_table_ptr make_empty_table()
+            {
+                node_table_ptr t(new tsk_node_table_t{},[](tsk_node_table_t* nt){delete nt;});
+                if (t.get() == NULL) {
+                    throw std::runtime_error("Out of memory");
+                }
+                return t;
+            }
+
+            tsk_node_table_t* allocate(tsk_node_table_t * the_table)
+            {
+                if (the_table != nullptr){ return the_table; }
+
+                node_table_ptr t(make_empty_table());
+                int ret = tsk_node_table_init(t.get(), 0);
+                // NOTE: check error is dangerous here.  If you don't
+                // use a smart pointer above, you will get a leak.
+                check_error(ret);
+                return t.release();
+            }
+
+            tsk_node_table_t*copy_construct_details(const NodeTable & other)
+            {
+                if(other.allocated_locally == false)
+                {
+                    // non-owning, so we share the pointer
+                    // to data
+                    return other.table;
+                }
+                node_table_ptr t(make_empty_table());
+                int ret = tsk_node_table_copy(other.table, t.get(), 0);
+                check_error(ret);
+                return t.release();
+            }
+
+        public:
+            explicit NodeTable(tsk_node_table_t *the_table) : table(allocate(the_table)), allocated_locally(the_table != table)
+            {
+                std::cout << "table constructor " << endl;
+            }
+
+            NodeTable(const NodeTable & other) : table(copy_construct_details(other)), allocated_locally(other.allocated_locally)
```

Comment:
So, I think copy should always be a deep copy here. We allocate a new node_table_t and call node_table_copy() on it. Does this allow us to simplify things a bit here?

Reply:
Yes, probably.  I was attempting to make this "copy-free":

```cpp
TableCollection foo;
auto n(foo.nodes);
```

If the constructor always makes a deep copy, then this will have the same effect:

```cpp
// Refer to the member table by reference
auto & n = foo.nodes;
```